src/connectedk8s/azext_connectedk8s/_utils.py

```diff
@@ -379,6 +399,85 @@ def is_guid(guid):
         return False
 
 
+def get_latest_extension_version(custom_logger, extension_name='connectedk8s', max_retries=3, timeout=2):
+    git_url = "https://raw.githubusercontent.com/Azure/azure-cli-extensions/master/src/index.json"
+    # or use "https://aka.ms/azure-cli-extension-index-v1"
+    try:
+        with requests.Session() as s:
+            s.mount(git_url, requests.adapters.HTTPAdapter(max_retries=max_retries))
+            response = s.get(git_url, timeout=timeout)
+            response_json = response.json()
+            version_list = []
+            for ver in response_json["extensions"][extension_name]:
+                version_list.append(ver["metadata"]["version"])
+            version_list.sort()
+            return version_list[-1]
+    except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as ex:
+        custom_logger.error('Internet connectivity problem detected. Error: {}'.format(str(ex)))
+    except Exception as ex:
+        custom_logger.error("Failed to get the latest connectedk8s version from '%s'. %s", git_url, str(ex))
+    return None
+
+
+def get_existing_extension_version(extension_name='connectedk8s'):
+    from azure.cli.core.extension import get_extensions
+    extensions = get_extensions()
+    if extensions:
+        for ext in extensions:
+            if ext.name == extension_name:
+                return ext.version or 'Unknown'
+
+    return 'NotFound'
+
+
+def check_connectivity(url='https://azure.microsoft.com', max_retries=5, timeout=1):
+    import timeit
+    start = timeit.default_timer()
+    success = None
+    try:
+        with requests.Session() as s:
+            s.mount(url, requests.adapters.HTTPAdapter(max_retries=max_retries))
+            s.head(url, timeout=timeout)
+            success = True
+    except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as ex:
+        logger.error('Internet connectivity problem detected. Error: {}'.format(str(ex)))
+        success = False
+    except Exception as ex:  # pylint: disable=broad-except
+        logger.error("Failed to check Internet connectivity. Error: %s", str(ex))
+        success = False
+    stop = timeit.default_timer()
+    logger.debug('Connectivity check: %s sec', stop - start)
+    return success
+
+
+def validate_azure_management_reachability(subscription_id):
+    try:
+        get_subscription_client().get(subscription_id)
+    except Exception as ex:
+        logger.warning("Not able to reach azure management endpoints. Exception: " + str(ex))
+
+
+def check_provider_registrations(cli_ctx):
+    try:
+        rp_client = _resource_providers_client(cli_ctx)
+        cc_registration_state = rp_client.get(consts.Connected_Cluster_Provider_Namespace).registration_state
+        if cc_registration_state != "Registered":
```

Comment:
It would be better to fail here if the state in not 'Registered'. Instead of waiting for cc creation step and encountering failure there. Collect exception telemetry also and set user fault